usefulUtilities/hiFiCalendar/entityServerScripts/tokenServer.js

```diff
@@ -0,0 +1,254 @@
+        });
+    };
+
+
+    // After the app is finished linking, it calls initialize rooms to refresh all of their tokens
+    this.initializeRooms = function(id, params) {
+        if (that.entityID === id) {
+
+            try {
+                that.userData = JSON.parse(Entities.getEntityProperties(that.entityID, ['userData']).userData);
+                that.roomConfig = that.userData.roomConfig = JSON.parse(params[5]);
+                that.makeCalendarToRoomAddressMap(that.roomConfig);
+            } catch (e){
+                console.log("problems parsing room config", e);
+            }
+
+            var privateUserData = {};
+
+            that.token = privateUserData.token = params[0];
+            that.refreshToken = privateUserData.refreshToken = params[1];
+            that.expireTime = that.userData.expireTime = params[2]; 
+            that.timezone = that.userData.timezone = params[3];
+            that.timezoneOffset = that.userData.timezoneOffset = params[4]; 
+
+            Entities.editEntity(that.entityID, {
+                userData: JSON.stringify(that.userData),
+                privateUserData: JSON.stringify(privateUserData)
+            });
+
+            that.calendarScheduleIDs.forEach(function(entity) {
+                if (that.calendarToRoomAddressMap[entity.id]) {
+                    Entities.callEntityMethod(entity.id, "refreshToken", [
+                        that.token, 
+                        that.expireTime, 
+                        that.timezoneOffset, 
+                        that.timezone,
+                        that.calendarToRoomAddressMap[entity.id]
+                    ]);
+                }
+            });
+        }
+    };
+
+
+    // This function checks to see if the token needs refreshing based on the token sent in the message
+    // If token sent is the same as the stored token, make a request to Google to refresh it.
+    this.tokenCheck = function(id, params) {
+        if (params[0] === that.token || (!params[0] && !that.token)) {
+            that.request({
+                uri: "https://highfidelity.co/hifiCalendar/api/request_token",
+                method: "POST",
+                json: true,
+                body: {
+                    refresh_token: that.refreshToken
+                }
+            }, function(error, response) {
+                if (error || response.status !== "success") {
+                    Messages.sendMessage(CHANNEL, JSON.stringify({
+                        type: "ERROR",
+                        entityName: that.entityProperties.name,
+                        errorMessage: error,
+                        actionAttempted: "Refreshing Oauth Token with Google"
+                    }));
+                    that.tokenStatus = false;
+                    return;
+                } else {
+                    that.expireTime = new Date().valueOf() + response.expires_in * MS_TO_SEC + that.userData.timezoneOffset * MS_TO_SEC * MIN_PER_HR * MIN_PER_HR;
```

Comment:
I'm not fully sure why this is necessary either.  One of those things I didn't touch because there seemed to have been a lot of work figuring these parts out before me.  I'm going to have to log what expires_in is actually giving and how that relates to the timezoneOffset.

Reply:
After looking through this and logging, I am very hesitant to change this at the moment because I don't fully understand what is happening between the tokenServer and roomCalendar, but it is passing the smoke test.  I'm making a note to revisit this if we are having problems with teams in other timezones.